### DISCUSSION TITLE
Empty output of jekyll-toc

### DISCUSSION THREAD

**Discussion by pavelsr** [allejo/jekyll-toc#31] [Q&A] [answered]
Please see sample project
https://github.com/pavelsr/jekyll-toc-test

1. `_includes/toc.html` exist and latest
2. `{% include toc.html html=content %}` called

But output is empty

```
$ jekyll --version
jekyll 4.0.0
```

What could be the problem?

**Comment by allejo** [comment]
To my knowledge, unless you manually create one, pages in Jekyll do not have a `content` variable; or if it does, it's probably not doing what you'd think it's doing. The `toc.html` include should be used inside [a layout](https://jekyllrb.com/docs/step-by-step/04-layouts/) where `content` does exist.

**Comment by pavelsr** [comment]
@allejo thanks for quick response.
I moved toc.html include to layout but still empty result
Please see new repo
https://github.com/pavelsr/jekyll-toc-html-test

**Comment by allejo** [comment]
A few things, see the output of `{{ content }}` in your `index.html`. There's nothing there. What you see, you are generating manually in the loop for `site.sections`.

Secondly, this TOC script will only work if your headings have `id` attributes. If you just have plain headings (e.g. `<h1>Hello</h1>`), this script won't work.

**Comment by allejo** [comment]
I recommend taking a look at how [collections in Jekyll](https://jekyllrb.com/docs/collections/) works because I *think* that's what you're trying to achieve?

**Comment by pavelsr** [comment]
> To my knowledge, unless you manually create one, pages in Jekyll do not have a content variable; or if it does, it's probably not doing what you'd think it's doing. The toc.html include should be used inside a layout where content does exist.

I found that it's not obligatory to place `{% include toc.html ...` in layout file, example below will work in regular page file placed somewhere in root of project. Just need to change from `html=content` to `html=page.content`

```html
---
---

<!doctype html>
<html>
  <head></head>
	<body>
		{% include toc.html html=page.content %}
		<h1 id="hello_world">Hello world</h1>
	</body>
</html>
```


> A few things, see the output of `{{ content }}` in your `index.html`. There's nothing there

You are right, output of `{{ content | inspect }}` in `index.html` is `nil`. But again, according [jekyll docs](https://jekyllrb.com/docs/variables/) correct notation is `page.content`

> Secondly, this TOC script will only work if your headings have id attributes. If you just have plain headings (e.g. `<h1>Hello</h1>`), this script won't work.

Yes, formally toc will not work since without heading id there will be no html anchor. 
But toc output will not empty, result `<ul><li>...</li></ul>` list just will be generated with `href="#"`, which points to top of page. Also seems like if you use `markdownify` [Luquid filter](https://jekyllrb.com/docs/liquid/filters/) heading id will be generated automatically (your [tests](https://raw.githubusercontent.com/allejo/jekyll-toc/master/_tests) are proving that)

As far as I understand main requirement for jekyll-toc is passing into `html` snippet [parameter](https://github.com/allejo/jekyll-toc#parameters) string with html code.

In my case output is totally empty, problem is that I can't make jekyll-toc work at all so I'm trying to understand why and that's why I created this issue (probably result of our discussion should be README.md improvement)

I called jekyll-toc same way like [docks.docker.com](https://github.com/docker/docker.github.io/blob/bbee70403d2974530eedf8f9b03f8235a63d281c/_layouts/docs.html#L268), [minimal-mistakes](https://github.com/mmistakes/minimal-mistakes/blob/f33158412536eee16d2dae63c5d646ebfd4d2142/_layouts/single.html#L41) or [bitcoin.org](https://github.com/bitcoin-dot-org/bitcoin.org/blob/6f40f9b1f0f70a57e1183b18b375f0389202a1a6/_layouts/post.html#L41) but for some reason output is empty.

> I recommend taking a look at how collections in Jekyll works because I think that's what you're trying to achieve?

I understood sense of collections, in my [test case](https://github.com/pavelsr/jekyll-toc-html-test) I just want try not keep all content in single markdown file but split content in some files, put them into `_sections` folder (so `sections` will be collection name) and generate single file using loop by `site.sections`.

**Comment by pavelsr** [comment, marked as answer]
Eureka, I found solution using `{% capture %}...{% endcapture %}` [tags](
https://shopify.dev/docs/liquid/reference/tags/variable-tags#capture) !

https://github.com/pavelsr/jekyll-toc-html-test/commit/bfc1f7b7c7f11660bdae97491b82a4f7440f9900

**Comment by allejo** [comment]
> I understood sense of collections, in my [test case](https://github.com/pavelsr/jekyll-toc-html-test) I just want try not keep all content in single markdown file but split content in some files, put them into `_sections` folder (so `sections` will be collection name) and generate single file using loop by `site.sections`.

Got it! That's what you were trying to do. Ok then yea, your finding of the `{% capture %}` tags would work nicely for you. Glad you were able to figure that out 😄
